### PR TITLE
Support ratis-shell base framework with info command

### PR DESCRIPTION
**Pull request by maobaolong**
- How to build

```console
mvn clean install  assembly:single -DskipTests -Dmaven.javadoc.skip=true -Dlicense.skip=true -Dcheckstyle.skip=true -Dfindbugs.skip=true -Denforcer.skip=true   -Djsse.enableSNIExtension=false  -Prelease -Papache-releas
```

- How to run
```
cd ratis/ratis-assembly/target/
tar -xzf apache-ratis-2.3.0-SNAPSHOT-bin.tar.gz
cd apache-ratis-2.3.0-SNAPSHOT
bin/ratis sh info -peers PEER0_HOST:PEER0_PORT,PEER1_HOST:PEER1_PORT,PEER2_HOST:PEER2_PORT
```

### CODE REVIEW

**Comment by maobaolong** [comment]
@szetszwo Would you please take a look at this PR, thank you.

**Comment by szetszwo** [comment]
@maobaolong , thanks for working on this.  I am looking at this pull request.

File: ratis-shell/pom.xml [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+                  </excludes>
+                </filter>
+              </filters>
+
+              <transformers>
+                <transformer implementation="org.apache.maven.plugins.shade.resource.ServicesResourceTransformer"/>
+
+                <transformer implementation="org.apache.maven.plugins.shade.resource.ManifestResourceTransformer">
+                  <manifestEntries>
+                    <Main-Class>org.apache.ratis.ratisshell.cli.sh.RatisShell</Main-Class>
```
Comment (szetszwo) [inline ratis-shell/pom.xml:121]: Let's use "org.apache.ratis.shell" for the package name.

File: ratis-shell/src/main/bin/ratis [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  done
+  "${JAVA}" -cp ${CLASSPATH} ${RATIS_SHELL_JAVA_OPTS} ${CLASS} ${PARAMETER} "${CLASS_ARGS[@]}"
+}
+
+function main {
+  LAUNCHER=
+  # If debugging is enabled propagate that through to sub-shells
+  if [[ $- == *x* ]]; then
+    LAUNCHER="bash -x"
+  fi
```
Comment (szetszwo) [inline ratis-shell/src/main/bin/ratis:47]: LAUNCHER is not used.  Should we remove it?

File: ratis-shell/pom.xml [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    </dependency>
+    <dependency>
+      <artifactId>ratis-metrics</artifactId>
+      <groupId>org.apache.ratis</groupId>
+    </dependency>
+    <dependency>
+      <groupId>org.reflections</groupId>
+      <artifactId>reflections</artifactId>
+      <version>0.9.12</version>
+    </dependency>
```
Comment (szetszwo) [inline ratis-shell/pom.xml:88]: Let's do not add any dependencies in this pull request.  We will add them later we they are used.

**Review by szetszwo — CHANGES_REQUESTED** [review]
@maobaolong , thanks a lot for your contribution!

This pull request is still quite big.  How about we only work on RATIS-1422 in this pull request?

I suggest to remove all the java files and add only the maven module and the shell scripts.

**Comment by szetszwo** [comment]
See https://issues.apache.org/jira/secure/attachment/13035814/525_review.patch for the review suggestions.

File: ratis-shell/pom.xml (hunk 3 above)
Comment (maobaolong) [reply ratis-shell/pom.xml:88]: Done

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Comment by szetszwo** [comment]
> Error:  src/main/java/org/apache/ratis/shell/cli/sh/RatisShell.java:[20,1] (design) HideUtilityClassConstructor: Utility classes should not have a public or default constructor.

The checkstyle warning above is a false positive.

**Pull request merged**